omnigibson/envs/data_wrapper.py

```diff
@@ -0,0 +1,525 @@
+            self.hdf5_file["data"].attrs["n_steps"] = self.step_count
+            self.hdf5_file.close()
+
+
+class DataCollectionWrapper(DataWrapper):
+    """
+    An OmniGibson environment wrapper for collecting data in an optimized way.
+
+    NOTE: This does NOT aggregate observations. Please use DataPlaybackWrapper to aggregate an observation
+    dataset!
+    """
+
+    def __init__(self, env, output_path, viewport_camera_path="/World/viewer_camera", only_successes=True):
+        """
+        Args:
+            env (Environment): The environment to wrap
+            output_path (str): path to store hdf5 data file
+            viewport_camera_path (str): prim path to the camera to use when rendering the main viewport during
+                data collection
+            only_successes (bool): Whether to only save successful episodes
+        """
+        # Store additional variables needed for optimized data collection
+        self.max_state_size = 0
+        self.current_transitions = dict()
+
+        # Add callbacks on import / remove objects and systems
+        og.sim.add_callback_on_system_init(
+            name="data_collection", callback=lambda system: self.add_transition_info(obj=system, add=True)
+        )
+        og.sim.add_callback_on_system_clear(
+            name="data_collection", callback=lambda system: self.add_transition_info(obj=system, add=False)
+        )
+        og.sim.add_callback_on_import_obj(
+            name="data_collection", callback=lambda obj: self.add_transition_info(obj=obj, add=True)
+        )
+        og.sim.add_callback_on_remove_obj(
+            name="data_collection", callback=lambda obj: self.add_transition_info(obj=obj, add=False)
+        )
+
+        # Disable all render products to save on speed
+        # See https://forums.developer.nvidia.com/t/speeding-up-simulation-2023-1-1/300072/6
+        for sensor in VisionSensor.SENSORS.values():
+            sensor.render_product.hydra_texture.set_updates_enabled(False)
+
+        # Set the main viewport camera path
+        og.sim.viewer_camera.active_camera_path = viewport_camera_path
+
+        # Use asynchronous rendering for faster performance
+        lazy.carb.settings.get_settings().set_bool("/app/asyncRendering", True)
+        lazy.carb.settings.get_settings().set_bool("/app/asyncRenderingLowLatency", True)
+
+        # Disable mouse grabbing since we're only using the UI passively
+        lazy.carb.settings.get_settings().set_bool("/physics/mouseInteractionEnabled", False)
+        lazy.carb.settings.get_settings().set_bool("/physics/mouseGrab", False)
+        lazy.carb.settings.get_settings().set_bool("/physics/forceGrab", False)
+        lazy.carb.settings.get_settings().set_bool("/physics/suppressReadback", True)
+
+        # Set the dump filter for better performance
+        env.scene.object_registry.set_dump_filter(dump_filter=lambda obj: obj.is_active)
```

Comment:
This whole is_active mechanism feels very very brittle and like it can bite us down the road, especially due to its reliance on the sleep mechanism. Let's at least leave a TODO here and open a ticket to follow up on this after we have full physics state saving/loading via custom views alongside the usual dump/load state mechanism.

Reply:
done!